cypress/e2e/awx/administration/wfApprovals.cy.ts

```diff
@@ -0,0 +1,245 @@
+import { Organization } from '../../../../frontend/awx/interfaces/Organization';
+import { Project } from '../../../../frontend/awx/interfaces/Project';
+import { AwxUser } from '../../../../frontend/awx/interfaces/User';
+// import { WorkflowApproval } from '../../../../frontend/awx/interfaces/WorkflowApproval';
+import { WorkflowJob } from '../../../../frontend/awx/interfaces/WorkflowJob';
+import { WorkflowJobTemplate } from '../../../../frontend/awx/interfaces/WorkflowJobTemplate';
+import { WorkflowNode } from '../../../../frontend/awx/interfaces/WorkflowNode';
+import { awxAPI } from '../../../support/formatApiPathForAwx';
+
+describe('Workflow Approvals Tests', () => {
+  let organization: Organization;
+  let project: Project;
+  let user: AwxUser;
+  let userWFApprove: AwxUser;
+  let userWFDeny: AwxUser;
+  let userWFCancel: AwxUser;
+  let inventory: Inventory;
+  let inventorySource: InventorySource;
+  let jobTemplate: JobTemplate;
+  let workflowJobTemplate: WorkflowJobTemplate;
+  let wfjobTemplateNode: WorkflowNode;
+  let approvalWFNode: WorkflowNode;
+  //   let jobName = '';
+
+  before(function () {
+    cy.awxLogin();
+  });
+
+  beforeEach(function () {
+    organization = this.globalOrganization as Organization;
+    cy.createAwxProject(
+      { organization: (this.globalOrganization as Organization).id },
+      'https://github.com/ansible/test-playbooks'
+    ).then((proj) => {
+      project = proj;
+
+      cy.createAwxUser(organization).then((u) => {
+        user = u;
+      });
+      cy.createAwxUser(organization).then((u) => {
+        userWFApprove = u;
+      });
+      cy.createAwxUser(organization).then((u) => {
+        userWFDeny = u;
+      });
+      cy.createAwxUser(organization).then((u) => {
+        userWFCancel = u;
+      });
+      cy.createAwxInventory({ organization: organization.id })
+        .then((i) => {
+          inventory = i;
+        })
+        .then(() => {
+          cy.createAwxInventorySource(inventory, project).then((invSrc) => {
+            inventorySource = invSrc;
+          });
+          cy.createAwxJobTemplate(
+            {
+              organization: organization.id,
+              project: project.id,
+              inventory: inventory.id,
+            },
+            'hello world.yml'
+          ).then((jt) => {
+            jobTemplate = jt;
+
+            cy.createAwxWorkflowJobTemplate({
+              organization: organization.id,
+              inventory: inventory.id,
+            }).then((wfjt) => {
+              workflowJobTemplate = wfjt;
+              cy.createAwxWorkflowVisualizerWJTNode(workflowJobTemplate).then((wfjtNode) => {
+                wfjobTemplateNode = wfjtNode;
+              });
+              cy.createAwxWorkflowVisualizerApprovalNode(workflowJobTemplate).then((appNode) => {
+                approvalWFNode = appNode;
+                cy.createWorkflowJTAlwaysNodeLink(approvalWFNode, wfjobTemplateNode);
```

Comment:
Potential problem, createAwxWorkflowVisualizerWJTNode and createAwxWorkflowVisualizerApprovalNode are running simulatenously (in parallel) and thus when createAwxWorkflowVisualizerApprovalNode is finished and you are creating link between them, you can't be sure the WJTNode is already finished and it may fail.

Easiest solution is to call them in nested style:
 cy.createAwxWorkflowJobTemplate({
              organization: organization.id,
              inventory: inventory.id,
            }).then((wfjt) => {
              workflowJobTemplate = wfjt;
              cy.createAwxWorkflowVisualizerWJTNode(workflowJobTemplate).then((wfjtNode) => {
                wfjobTemplateNode = wfjtNode;
                // this ensures next node is called after first node
                  cy.createAwxWorkflowVisualizerApprovalNode(workflowJobTemplate).then((appNode) => {
                approvalWFNode = appNode;
                // and here we are sure both nodes are created
                cy.createWorkflowJTAlwaysNodeLink(approvalWFNode, wfjobTemplateNode);
              });
            
            If you want to go fancy, you can try to use cy.all (as google says, never used it before, you must double check), which can wait for parallel execution for both nodes, this will speed up tests a bit.